File: 29Items-Projects/7Algorithms/4-Greedy-algorithms/java/src/main/java/com/example/greedy/algorithms/IntervalScheduling.java
package com.example.greedy.algorithms;

import com.example.greedy.model.Interval;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public final class IntervalScheduling {
    private IntervalScheduling() {
    }

    public static List<Interval> schedule(List<Interval> intervals) {
        for (Interval interval : intervals) {
            if (interval.finish() < interval.start()) {
                throw new IllegalArgumentException(interval.name() + " finishes before it starts");
            }
        }

        List<Interval> selected = new ArrayList<>();
        int lastFinish = -1;
        for (Interval interval : intervals.stream()
                .sorted(Comparator.comparingInt(Interval::finish)
                        .thenComparingInt(Interval::start)
                        .thenComparing(Interval::name))
                .collect(Collectors.toList())) {
            if (interval.start() >= lastFinish) {
                selected.add(interval);
                lastFinish = interval.finish();
            }
        }
        return selected;
    }
}

File: 29Items-Projects/7Algorithms/4-Greedy-algorithms/python/greedy_algorithms/algorithms/interval_scheduling.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Interval:
    name: str
    start: int
    finish: int


def schedule_intervals(intervals: list[Interval]) -> list[Interval]:
    for interval in intervals:
        if interval.finish < interval.start:
            raise ValueError(f"{interval.name} finishes before it starts")

    result: list[Interval] = []
    last_finish = -1
    for interval in sorted(intervals, key=lambda item: (item.finish, item.start, item.name)):
        if interval.start >= last_finish:
            result.append(interval)
            last_finish = interval.finish
    return result
